### PR TITLE
Fix: can't apply or remove a coupon whose code is "0".

### LINKED ISSUE #35210: 0 (zero) in apply coupon returns 'empty coupon code'
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [ ] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

It allows to create discount code: 0
but when trying to apply '0' coupon in cart, it returns 

'Coupon "0" does not exist!'

possibly because for validating you have used PHP 'empty' function which returns true for 0
https://www.php.net/manual/en/function.empty.php#usernotes

### Expected behavior

It should have applied the coupon in the cart.

### Actual behavior

it returns 'Coupon "0" does not exist!' when trying to apply coupon code: 0 (zero)

### Steps to reproduce

1. Create coupon code: 0
2. and try to apply it on cart

### WordPress Environment

`
### WordPress Environment ###

WordPress address (URL): http://localhost/wc_cart
Site address (URL): http://localhost/wc_cart
WC Version: 6.6.1
REST API Version: ✔ 6.6.1
WC Blocks Version: ✔ 7.6.2
Action Scheduler Version: ✔ 3.4.0
Log Directory Writable: ✔
WP Version: 6.0.3
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.48 (Win64) OpenSSL/1.1.1k PHP/8.0.8
PHP Version: 8.0.8
PHP Post Max Size: 40 MB
PHP Time Limit: 120
PHP Max Input Vars: 1000
cURL Version: 7.76.1
OpenSSL/1.1.1k

SUHOSIN Installed: –
MySQL Version: 5.5.5-10.4.20-MariaDB
Max Upload Size: 40 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ❌ Your server does not have the SoapClient class enabled - some gateway plugins which use SOAP may not work as expected.
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 6.6.1
WC Database Prefix: wp_
Total Database Size: 4.61MB
Database Data Size: 3.14MB
Database Index Size: 1.47MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 2.06MB + Index: 0.03MB + Engine InnoDB
wp_postmeta: Data: 0.13MB + Index: 0.06MB + Engine InnoDB
wp_posts: Data: 0.06MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.05MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 23
page: 7
post: 2
product: 18
product_variation: 7
shop_coupon: 3

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (3) ###

Loco Translate: by Tim Whitlock – 2.6.2
Cart Drawer: by dev– 2.0
WooCommerce: by Automattic – 6.6.1 (update to version 7.0.0 is available)

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔

### WC Pages ###

Shop base: #6 - /shop/
Cart: #7 - /cart/
Checkout: #8 - /checkout/
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: GeneratePress
Version: 3.1.0 (update to version 3.1.3 is available)
Author URL: https://tomusborne.com
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
coupons
customer-effort-score-tracks
experimental-products-task
experimental-import-products-task
experimental-fashion-sample-products
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
wc-pay-promotion
wc-pay-welcome-page
wc-pay-subscriptions-page

Disabled Features: minified-js
settings

Daily Cron: ✔ Next scheduled: 2022-10-21 09:24:32 +00:00
Options: ✔
Notes: 54
Onboarding: skipped

### Action Scheduler ###

Complete: 4
Oldest: 2022-10-18 20:15:13 +0000
Newest: 2022-10-20 13:19:58 +0000

Pending: 1
Oldest: 2022-10-21 13:18:50 +0000
Newest: 2022-10-21 13:18:50 +0000

### Status report information ###

Generated at: 2022-10-20 13:55:38 +00:00
`

### Isolating the problem

- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by Konamiman**
### All Submissions:

-   [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Fixes a bug that allowed to create a coupon with code "0" but prevented if from being actually used. With the fix a coupon with code "0" can be used as any other coupon.

Also adds two new methods to the `StringUtil` class: `is_null_or_empty` and `is_null_or_whitespace`.

Closes #35210.

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

### How to test the changes in this Pull Request:

- [ ] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

1. Go to Marketing - Coupons and click "Add coupon".
2. Give the coupon the code `0` and an amount of 1 (Coupon data - General - Coupon amount), leave it as type "Fixed cart discount" and save it.
3. Go to the shop and add any product to the cart, then go to the cart.
4. Type `0 ` in "Coupon code", then click "Apply coupon". The coupon should be applied successfully and appear listed in "Cart totals".
5. Click the "Remove" link next to the coupon amount in "Cart totals", the coupon should be successfully removed.
6. Click "Proceed to checkout" and click on "Have a coupon? Click here to enter your code".
7. Enter the coupon code `0` and click "Apply coupon", the coupon should be applied successfully and appear listed in "Your order".
8. Click the "Remove" link next to the coupon amount in "Your order", the coupon should be successfully removed.
9. Apply the coupon again and complete the order.
10. Go to Marketing - Coupons in the admin area and verify that the coupon usage has been increased.
11. Go to the admin area and open the order details (WooCommerce - Orders). Verify that the coupon is listed under "Coupon(s)" and that its value is discounted from the order total.
12. Remove the coupon by clicking the "x" in the coupon name, verify that the operation succeeds and the coupon value is no longer discounted from the order total.
13. Go to Marketing - Coupons in the admin area and verify that the coupon usage has been decreased.
14. Add the coupon again by clickin "Add coupon", verify that the operation succeeds and the coupon value is discounted again from the order total.
15. Edit the coupon so that it has a maximum allowed usage (Usage limits - Usage limit per coupon) equal to the current usage count.
16. Try to apply the coupon again to a different order, you should get an error now.

### Testing with REST API

Verify that the coupon "0" can be created and applied via REST API too.

Example request to create (try `v1`, `v2` and `v3`):

```
curl --request POST \
  --url http://localhost/wp-json/wc/v3/coupons \
  --header 'Authorization: Basic {your app password}' \
  --header 'Content-Type: multipart/form-data' \
  --form code=0 \
  --form discount_type=percent \
  --form amount=1
```

Example request to apply to a new order:

```
curl --request POST \
  --url http://localhost/wp-json/wc/v3/orders \
  --header 'Authorization: Basic {your app password}' \
  --header 'Content-Type: application/json' \
  --data '{
	"line_items": [
    {
      "product_id": 95,
      "quantity": 1
    }
  ],
	"coupon_lines": [
		{
			"code": "0"
		}
	],
	"payment_method": "cod"
}'
```

Also verify that a `GET` request to the coupons endpoint with `code=0`  returns just the coupon with code `0` instead of all the existing coupons:

```
curl --request GET --url 'http://localhost/wp-json/wc/v3/coupons?code=0' --header 'Authorization: Basic {your app password}'
```

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?
-   [ ] Have you included testing instructions?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce/tests/php/src/Utilities/StringUtilTest.php [hunk 1]
```diff
@@ -74,4 +74,40 @@ public function test_plugin_name_from_plugin_file() {
...
+
+	/**
+	 * @testDox 'is_null_or_empty' should return true only if the value is null or an empty string.
+	 *
+	 * @testWith [null, true]
+	 *           ["", true]
+	 *           ["  ", false]
+	 *           ["0", false]
+	 *           ["foo", false]
+	 *           ["  foo  ", false]
```
Comment (coreymckrill) [inline plugins/woocommerce/tests/php/src/Utilities/StringUtilTest.php:86]: Cool, I didn't know you could put the test data right in the doc block!

**Review by coreymckrill — CHANGES_REQUESTED** [review]
Great testing instructions, works as described 👍 However, the problem still exists when working with coupons via the REST API:

Creating a coupon

```
curl --request POST \
  --url http://localhost:8888/wp-json/wc/v3/coupons \
  --header 'Authorization: Basic {your app password}' \
  --header 'Content-Type: multipart/form-data' \
  --form code=0 \
  --form discount_type=percent \
  --form amount=1
```

Returns 

```
{
	"code": "woocommerce_rest_empty_coupon_code",
	"message": "The coupon code cannot be empty.",
	"data": {
		"status": 400
	}
}
```

I think the problem is on [this line](https://github.com/woocommerce/woocommerce/blob/756fe32c048a6c8e4e81d33147398340e93adc84/plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-coupons-v2-controller.php#L271).

Applying a coupon while creating an order:

```
curl --request POST \
  --url http://localhost:8888/wp-json/wc/v3/orders \
  --header 'Authorization: Basic {your app password}' \
  --header 'Content-Type: application/json' \
  --data '{
	"line_items": [
    {
      "product_id": 95,
      "quantity": 1
    }
  ],
	"coupon_lines": [
		{
			"code": "0"
		}
	],
	"payment_method": "cod"
}'
```

Returns

```
{
	"code": "woocommerce_rest_invalid_coupon",
	"message": "Coupon code is required.",
	"data": {
		"status": 400
	}
}
```

It's not immediately obvious where the culprit is here. [This line](https://github.com/woocommerce/woocommerce/blob/756fe32c048a6c8e4e81d33147398340e93adc84/plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-orders-v2-controller.php#L670) seems to only check for null values, which `0` is not when using `is_null`, so I'm not sure where the problem is.

File: plugins/woocommerce/tests/php/src/Utilities/StringUtilTest.php (hunk 1 above)
Comment (Konamiman) [reply plugins/woocommerce/tests/php/src/Utilities/StringUtilTest.php:86]: [You are not the first one that learns this by reviewing my PRs](https://github.com/woocommerce/woocommerce/pull/36891#discussion_r1113491860). Apparently I'm the only one around here that reads manuals. 😄

**Comment by Konamiman** [comment]
@coreymckrill Not only was the bug still present in REST API, also the usage count wasn't being increased for "0" coupons! I've pushed a commit that fixed both things.

What I have noticed, though, is that coupon usage count is not being increased or checked when applying it to a new order via REST API v1 and v2; but this seems a different bug as it happens with all coupons, not only those with code "0". Or am I missing something?

File: plugins/woocommerce/includes/rest-api/Controllers/Version1/class-wc-rest-coupons-v1-controller.php [hunk 2]
```diff
@@ -10,6 +10,9 @@
  * @since    3.0.0
  */
 
+use Automattic\WooCommerce\Utilities\ArrayUtil;
```
Comment (coreymckrill) [inline plugins/woocommerce/includes/rest-api/Controllers/Version1/class-wc-rest-coupons-v1-controller.php]: `ArrayUtil` seems unused in this file?

File: plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-coupons-v2-controller.php [hunk 3]
```diff
@@ -8,6 +8,9 @@
  * @since   2.6.0
  */
 
+use Automattic\WooCommerce\Utilities\ArrayUtil;
```
Comment (coreymckrill) [inline plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-coupons-v2-controller.php]: Also `ArrayUtil` seems unused in this file?

**Review by coreymckrill — CHANGES_REQUESTED** [review]
👍 Creating coupon and applying coupon to order now works in the REST API. Also confirmed that max allowed usage is not violated (v3). However, I found one other thing that still doesn't work (sorry I didn't notice this before):

```
curl --request GET \
  --url 'http://localhost:8888/wp-json/wc/v3/coupons?code=0' \
  --header 'Authorization: Basic {your app password}' \
  --header 'Content-Type: multipart/form-data'
```

[v1](https://github.com/woocommerce/woocommerce/blob/e5d8b613734d81fa251b732021e7eb21fdb6365b/plugins/woocommerce/includes/rest-api/Controllers/Version1/class-wc-rest-coupons-v1-controller.php#L134), [v2](https://github.com/woocommerce/woocommerce/blob/e5d8b613734d81fa251b732021e7eb21fdb6365b/plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-coupons-v2-controller.php#L236), [v3](https://github.com/woocommerce/woocommerce/blob/e5d8b613734d81fa251b732021e7eb21fdb6365b/plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-coupons-controller.php#L40)

> What I have noticed, though, is that coupon usage count is not being increased or checked when applying it to a new order via REST API v1 and v2; but this seems a different bug as it happens with all coupons, not only those with code "0". 

I noticed this too, and I think it's a separate bug. It does not appear to have an existing issue created for it. However, since it only affects legacy versions of the API, is it worth fixing at this point?

**Comment by Konamiman** [comment]
Pushed a commit that fixed the `GET` request and also a nasty bug I had introduced in `apply_coupon`, detected thanks to the almighty unit tests.

>However, since it only affects legacy versions of the API, is it worth fixing at this point?

Maybe. The bug is preventing the merchant from getting accurate coupon usage and is also allowing to apply coupons that shouldn't be applied, so I'd say it's severe enough to deserve attention.

**Review by coreymckrill — APPROVED** [review]

**Pull request merged**